serialize/encoder.go

```diff
@@ -0,0 +1,177 @@
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package serialize
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+
+	"github.com/m3db/m3x/checked"
+	"github.com/m3db/m3x/ident"
+)
+
+/*
+ * Serialization scheme to combat Thrift's allocation hell.
+ *
+ * Given Tags (i.e. key-values) this allows the bijective serialization to,
+ * and from Tags <--> []byte.
+ *
+ * Consider example, Tags: {"abc": "defg", "x": "foo"}
+ * this translates to:
+ * []byte(
+ *    MAGIC_MARKER + NUMBER_TAGS
+ *                 + LENGTH([]byte("abc"))  + []byte("abc")
+ *                 + LENGTH([]byte("defg")) + []byte("abc")
+ *                 + LENGTH([]byte("x"))    + []byte("x")
+ *                 + LENGTH([]byte("foo"))  + []byte("foo")
+ * )
+ *
+ * Where MAGIC_MARKER/NUMBER_TAGS/LENGTH are maximum 2 bytes.
+ */
+
+var (
+	byteOrder        binary.ByteOrder = binary.LittleEndian
+	headerMagicBytes                  = encodeUInt16(HeaderMagicNumber)
+)
+
+var (
+	errEncoderInUse      = errors.New("encoder already in use")
+	errTagLiteralTooLong = errors.New("literal is too long")
+)
+
+type encoder struct {
+	buf          *bytes.Buffer
+	checkedBytes checked.Bytes
+
+	pool EncoderPool
+}
+
+func newEncoder(initialSize int, pool EncoderPool) Encoder {
+	b := make([]byte, 0, initialSize)
+	cb := checked.NewBytes(nil, nil)
+	return &encoder{
+		buf:          bytes.NewBuffer(b),
+		checkedBytes: cb,
+		pool:         pool,
+	}
+}
+
+func (e *encoder) Encode(srcTags ident.TagIterator) error {
+	if e.checkedBytes.NumRef() > 0 {
+		return errEncoderInUse
+	}
+
+	tags := srcTags.Duplicate()
+	defer tags.Close()
+
+	numTags := tags.Remaining()
+	if numTags > MaxNumberTags {
+		return fmt.Errorf("too many tags to encode (%d), limit is: %d", numTags, MaxNumberTags)
+	}
+
+	if _, err := e.buf.Write(headerMagicBytes); err != nil {
+		e.buf.Reset()
+		return err
+	}
+
+	if _, err := e.buf.Write(encodeUInt16(uint16(numTags))); err != nil {
+		e.buf.Reset()
+		return err
+	}
+
+	for tags.Next() {
+		tag := tags.Current()
+		if err := e.encodeTag(tag); err != nil {
+			e.buf.Reset()
+			return err
+		}
+	}
+
+	if err := tags.Err(); err != nil {
+		e.buf.Reset()
+		return err
+	}
+
+	e.checkedBytes.IncRef()
+	e.checkedBytes.Reset(e.buf.Bytes())
+
+	return nil
+}
+
+func (e *encoder) Data() checked.Bytes {
+	return e.checkedBytes
+}
+
+func (e *encoder) Reset() {
+	if e.checkedBytes.NumRef() == 0 {
```

Comment:
Hm, strange?  Should we also check that NumRef == 0 after final DecRef if we fall through?